File: leetcode-js/113-path-sum-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function(root, sum) {
    if (!root) {
        return [];
    }

    const result = [];
    _pathSum(root, sum, [], result);
    return result;
};

function _pathSum(node, sum, paths, result) {
    if (node.left) {
        _pathSum(node.left, sum - node.val, paths.concat([node.val]), result);
    }

    if (node.right) {
        _pathSum(node.right, sum - node.val, paths.concat([node.val]), result);
    }

    if (!node.left && !node.right && sum === node.val) {
        result.push(paths.concat([node.val]));
    }
}

File: leetcode-java/113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> cur = new ArrayList<Integer>();
        helper(res, cur, root, sum);
        return res;
    }

    void helper(List<List<Integer>> res, List<Integer> cur, TreeNode node, int sum) {
        if (node == null) {
            return;
        }
        cur.add(node.val);
        if (node.left == null && node.right == null && node.val == sum) {
            res.add(new ArrayList<Integer>(cur));
        } else {
            helper(res, cur, node.left, sum - node.val);
            helper(res, cur, node.right, sum - node.val);
        }
        cur.remove(cur.size() - 1);
    }
}
